File: 2015_rust/src/bin/day_2.rs
const INPUT: &str = include_str!("../../input/day_2.txt");

fn parse_input() -> Vec<Vec<u32>> {
    INPUT
        .lines()
        .map(|line| line.split('x').map(|n| n.parse::<u32>().unwrap()).collect())
        .collect()
}

fn part1() -> u32 {
    parse_input()
        .iter()
        .map(|sides| {
            let a = sides[0] * sides[1];
            let b = sides[1] * sides[2];
            let c = sides[0] * sides[2];
            2 * (a + b + c) + a.min(b).min(c)
        })
        .sum()
}

fn part2() -> u32 {
    parse_input()
        .iter_mut()
        .map(|sides| {
            sides.sort();
            2 * (sides[0] + sides[1]) + sides.iter().product::<u32>()
        })
        .sum()
}

fn main() {
    println!("Part 1: {}", part1());
    println!("Part 2: {}", part2());
}

File: 2023_haskell/Day2.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}

type Bag = [Int]

main :: IO ()
main = do
    input <- map parse . lines <$> readFile "input/day_2.txt"
    print $ part1 input
    print $ part2 input


parse :: String -> [Bag]
parse = map parseBag . split ';' . drop 2 . dropWhile (/= ':')

parseBag :: String -> Bag
parseBag = foldl1 (zipWith max) . map (makeBag . words) . split ','

makeBag :: [String] -> Bag
makeBag [n, "red"] = [read n, 0, 0]
makeBag [n, "green"] = [0, read n, 0]
makeBag [n, "blue"] = [0, 0, read n]


part1 :: [[Bag]] -> Int
part1 = sum . map fst . filter (all (and . zipWith (>=) [12, 13, 14]) . snd) . zip [1..]

part2 :: [[Bag]] -> Int
part2 = sum . map (product . foldl1 (zipWith max))


split :: Eq a => a -> [a] -> [[a]]
split sep xs = case break (== sep) xs of
    (_, []) -> [xs]
    (ls, _:rs) -> ls : split sep rs
